[mlir][llvm dialect] Verify element type of nested types

newling opened the pull request:
Before this PR, this was valid
```
 %0 = llvm.mlir.constant(dense<[1, 2]> : vector<2xi32>) : vector<2xf32>
```

but this was not:

```
%0 = llvm.mlir.constant(1 : i32) : f32
```

because only scalar types were checked for compatibility, not the element types of nested types. 


Another additional check that this PR adds is to verify the float semantics. Before this PR, 

```
 %cst = llvm.mlir.constant(1.0 : bf16) : f16
 ```
 
 was considered valid (because bf16 and f16 both have 16 bits), but with this PR it is not considered valid. 
 
 This PR also moves all tests on the verifier of the llvm constant op into a single file. 
 
The total number of tests with this PR increases by 3. 

To summarize the state after this PR. Valid:

Invalid:
```
%0 = llvm.mlir.constant(dense<[128, 1024]> : vector<2xi32>) : vector<2xf32>
%0 = llvm.mlir.constant(dense<[128., 1024.]> : vector<2xbf16>) : vector<2xf16>
```

Valid:
```
%0 = llvm.mlir.constant(dense<[128., 1024.]> : vector<2xf32>) : vector<2xi32>
%0 = llvm.mlir.constant(dense<[128, 1024]> : vector<2xi64>) : vector<2xi8>
```

and identical valid/invalid cases for the scalar cases.

newling committed:
move them to the correct place

newling committed:
add checks on element attributes for types

gysit commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -3112,6 +3113,18 @@ static int64_t getNumElements(Type t) {
+    type = arrayType.getElementType();
+  if (auto vecType = dyn_cast<VectorType>(type))
+    return vecType.getElementType();
+  if (auto tenType = dyn_cast<TensorType>(type))
```
nit: the TensorType can be dropped. It is not a compatible LLVM type (see `isCompatibleOuterType`) and since ConstantOp returns an LLVM type there should be no tensors here.

newling replied:
Should the following fail to verify? 

```
  %0 = llvm.mlir.constant(dense<[1.0, 1.0]> : tensor<2xf64>) : vector<2xf64>
```

It currently roundtrips without issue. 

Maybe what is confusing here is that I'm using this method to test both the attribute's element type as well the constant op's result's element type?

gysit commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -3242,24 +3246,42 @@ LogicalResult LLVM::ConstantOp::verify() {
+    }
+    unsigned floatWidth = APFloat::getSizeInBits(attributeFloatSemantics);
+    if (isa<IntegerType>(constantElementType)) {
+      if (!constantElementType.isInteger(floatWidth)) {
+        return emitOpError() << "expected integer type of width " << floatWidth;
+      }
```
```suggestion
      if (!constantElementType.isInteger(floatWidth)) 
        return emitOpError() << "expected integer type of width " << floatWidth;
```
nit: for single line ifs we usually drop the braces

gysit commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -3271,13 +3293,23 @@ LogicalResult LLVM::ConstantOp::verify() {
+
+    Type attrElmType = getElementType(elementsAttr.getType());
+    Type resultElmType = getElementType(getType());
+    if (auto floatType = dyn_cast<FloatType>(attrElmType)) {
+      return verifyFloatSemantics(floatType.getFloatSemantics(), resultElmType);
+    }
```
```suggestion
    if (auto floatType = dyn_cast<FloatType>(attrElmType)) 
      return verifyFloatSemantics(floatType.getFloatSemantics(), resultElmType);
```
nit: Same as above

gysit commented on mlir/test/Dialect/LLVMIR/invalid.mlir:
```diff
@@ -439,6 +439,105 @@ llvm.func @scalable_vec_requires_splat() -> vector<[4]xf64> {
+  llvm.return %0 : !llvm.struct<(f64, f64)>
+}
+
+// -----
```
```suggestion
```
nit: one split can be dropped

gysit commented on mlir/test/Dialect/LLVMIR/invalid.mlir:
```diff
@@ -439,6 +439,105 @@ llvm.func @scalable_vec_requires_splat() -> vector<[4]xf64> {
+// -----
+
+llvm.func @int_attr_requires_int_type() -> f32 {
+  // expected-error @+1 {{expected integer type}}
```
```suggestion
  // expected-error @below {{expected integer type}}
```
ulta nit: can you use below everywhere in the moved tests? It is preferred over @+1

gysit reviewed (commented):
Thanks a lot for improving the verifiers!

I have added some small nit comments but otherwise looks good to go.

newling committed:
address a few review comments

newling committed:
use original error message (as per review comment)

newling commented:
@gysit Thank you for reviewing this! I think I have addressed your comments. The one I'm not sure about is the removal of TensorType from the new static function, `getElementType`. Please see my inline comment.

gysit reviewed (approved):
Thanks for addressing the comment!

LGTM modulo some leftover nit.

I did one more search and realized that the tensor type is also used for llvm.array constants. That is actually a somewhat justified use case. Let's keep your current version of the code then!

newling committed:
remove one liner if statement's { and } as per review comment

newling commented:
> Thanks for addressing the comment!
> 
> LGTM modulo some leftover nit.
> 
> I did one more search and realized that the tensor type is also used for llvm.array constants. That is actually a somewhat justified use case. Let's keep your current version of the code then!

Thanks for investigating this further, and thanks again for the review!

If there are no more review comments on this I'll land it on/after Monday.
